src/java/htsjdk/samtools/util/BlockCompressedInputStream.java

```diff
@@ -354,53 +366,76 @@ private static boolean isValidBlockHeader(final byte[] buffer) {
+        mCurrentOffset = 0;
+        if (mCurrentBlock.mException != null) {
+            if (mCurrentBlock.mException instanceof IOException) {
+                throw (IOException)mCurrentBlock.mException;
+            } else if (mCurrentBlock.mException instanceof RuntimeException) {
+                throw (RuntimeException)mCurrentBlock.mException;
+            } else {
+                throw new RuntimeException(mCurrentBlock.mException);
+            }
         }
-        final int blockLength = unpackInt16(mFileBuffer, BlockCompressedStreamConstants.BLOCK_LENGTH_OFFSET) + 1;
-        if (blockLength < BlockCompressedStreamConstants.BLOCK_HEADER_LENGTH || blockLength > mFileBuffer.length) {
-            throw new IOException("Unexpected compressed block length: " + blockLength);
+    }
+    
+    private DecompressedBlock nextBlock() {
+        if (mFileBuffer == null) {
+            mFileBuffer = new byte[BlockCompressedStreamConstants.MAX_COMPRESSED_BLOCK_SIZE];
         }
-        final int remaining = blockLength - BlockCompressedStreamConstants.BLOCK_HEADER_LENGTH;
-        count = readBytes(mFileBuffer, BlockCompressedStreamConstants.BLOCK_HEADER_LENGTH, remaining);
-        if (count != remaining) {
-            throw new FileTruncatedException("Premature end of file");
+        long blockAddress = mStreamOffset;
+        try {
+            final int headerByteCount = readBytes(mFileBuffer, 0, BlockCompressedStreamConstants.BLOCK_HEADER_LENGTH);
+            mStreamOffset += headerByteCount;
+            if (headerByteCount == 0) {
+                // Handle case where there is no empty gzip block at end.
+                return new DecompressedBlock(blockAddress, new byte[0], 0);
+            }
+            if (headerByteCount != BlockCompressedStreamConstants.BLOCK_HEADER_LENGTH) {
+                return new DecompressedBlock(blockAddress, headerByteCount, new IOException("Premature end of file"));
+            }
+            final int blockLength = unpackInt16(mFileBuffer, BlockCompressedStreamConstants.BLOCK_LENGTH_OFFSET) + 1;
+            if (blockLength < BlockCompressedStreamConstants.BLOCK_HEADER_LENGTH || blockLength > mFileBuffer.length) {
+                return new DecompressedBlock(blockAddress, blockLength, new IOException("Unexpected compressed block length: " + blockLength));
+            }
+            final int remaining = blockLength - BlockCompressedStreamConstants.BLOCK_HEADER_LENGTH;
+            final int dataByteCount = readBytes(mFileBuffer, BlockCompressedStreamConstants.BLOCK_HEADER_LENGTH, remaining);
+            mStreamOffset += dataByteCount;
+            if (dataByteCount != remaining) {
+                return new DecompressedBlock(blockAddress, blockLength, new FileTruncatedException("Premature end of file"));
+            }
+            final byte[] decompressed = inflateBlock(mFileBuffer, blockLength, !Defaults.USE_ASYNC_IO && mCurrentBlock != null ? mCurrentBlock.mBlock : null);
```

Comment:
also, add a comment re: why the buffer cannot be reused in the multithreading context